syscall_sdl.c

```diff
@@ -9,9 +9,75 @@
 
 #include "state.h"
 
+/* For optimization, the capcity of event queues must be the power of two to
+ * avoid the expensive modulo operation, the details are explained here:
+ * https://stackoverflow.com/questions/10527581/why-must-a-ring-buffer-size-be-a-power-of-2
+ */
+#define EVENT_QUEUE_CAPACITY 128
+
+enum {
+    KEY_EVENT = 0,
+    MOUSE_MOTION_EVENT = 1,
+    MOUSE_BUTTON_EVENT = 2,
+};
+
+typedef struct {
+    uint32_t keycode;
+    uint8_t state;
+} key_event_t;
+
+typedef struct {
+    int32_t xrel, yrel;
+} mouse_motion_t;
+
+typedef struct {
+    uint8_t button;
+    uint8_t state;
+} mouse_button_t;
+
+typedef struct {
+    uint32_t type;
+    union {
+        key_event_t key_event;
+        mouse_motion_t mouse_motion;
```

Comment:
How about wrapping mouse in another struct? That is,
```c
struct mouse {
    mouse_motion_t motion;
    mouse_button_t button;
}
```

Then, we can refer to the members with prefix `mouse`.

Reply:
How about a nested named union instance? A single `struct` containing all mouse events can take up unnecessary space.

Reply:
> How about a nested named union instance? A single `struct` containing all mouse events can take up unnecessary space.

It sounds good.